frame/fast-unstake/src/lib.rs

```diff
@@ -0,0 +1,500 @@
+	use frame_election_provider_support::ElectionProvider;
+	use frame_support::pallet_prelude::*;
+	use frame_system::{pallet_prelude::*, RawOrigin};
+	use pallet_nomination_pools::PoolId;
+	use pallet_staking::Pallet as Staking;
+	use sp_runtime::{
+		traits::{Saturating, Zero},
+		DispatchResult,
+	};
+	use sp_staking::EraIndex;
+	use sp_std::{prelude::*, vec::Vec};
+	use weights::WeightInfo;
+
+	#[derive(scale_info::TypeInfo, codec::Encode, codec::Decode, codec::MaxEncodedLen)]
+	#[codec(mel_bound(T: Config))]
+	#[scale_info(skip_type_params(T))]
+	pub struct MaxChecked<T: Config>(sp_std::marker::PhantomData<T>);
+	impl<T: Config> frame_support::traits::Get<u32> for MaxChecked<T> {
+		fn get() -> u32 {
+			<T as pallet_staking::Config>::BondingDuration::get() + 1
+		}
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config:
+		frame_system::Config
+		+ pallet_staking::Config<
+			CurrencyBalance = <Self as pallet_nomination_pools::Config>::CurrencyBalance,
+		> + pallet_nomination_pools::Config
+	{
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>>
+			+ IsType<<Self as frame_system::Config>::RuntimeEvent>
+			+ TryInto<Event<Self>>;
+
+		/// The amount of balance slashed per each era that was wastefully checked.
+		///
+		/// A reasonable value could be `runtime_weight_to_fee(weight_per_era_check)`.
+		type SlashPerEra: Get<BalanceOf<Self>>;
+
+		/// The origin that can control this pallet.
+		type ControlOrigin: frame_support::traits::EnsureOrigin<Self::Origin>;
+
+		/// The weight information of this pallet.
+		type WeightInfo: WeightInfo;
+	}
+
+	/// The current "head of the queue" being unstaked.
+	#[pallet::storage]
+	pub type Head<T: Config> =
+		StorageValue<_, UnstakeRequest<T::AccountId, MaxChecked<T>>, OptionQuery>;
+
+	/// The map of all accounts wishing to be unstaked.
+	///
+	/// Points the `AccountId` wishing to unstake to the optional `PoolId` they wish to join
+	/// thereafter.
+	#[pallet::storage]
+	pub type Queue<T: Config> = CountedStorageMap<_, Twox64Concat, T::AccountId, Option<PoolId>>;
+
+	/// Number of eras to check per block.
+	///
+	/// If set to 0, this pallet does absolutely nothing.
+	///
+	/// Based on the amount of weight available at `on_idle`, up to this many eras of a single
+	/// nominator might be checked.
+	#[pallet::storage]
+	pub type ErasToCheckPerBlock<T: Config> = StorageValue<_, u32, ValueQuery>;
+
+	/// The events of this pallet.
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A staker was unstaked.
+		Unstaked { stash: T::AccountId, maybe_pool_id: Option<PoolId>, result: DispatchResult },
+		/// A staker was slashed for requesting fast-unstake whilst being exposed.
+		Slashed { stash: T::AccountId, amount: BalanceOf<T> },
+		/// A staker was partially checked for the given eras, but the process did not finish.
+		Checked { stash: T::AccountId, eras: Vec<EraIndex> },
```

Comment:
This seems confusing, if a staker is still being checked (not finished), would it be better to name this event something like `BeingChecked`, `Checking` or something else that implies that the process is not finished?

Reply:
I agree, confusing, `Checking` is better 👍